### DISCUSSION TITLE
Invalid character in header content authorization

### DISCUSSION THREAD

**Discussion by tfendt-belong-gg** [renovatebot/renovate#11225] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### Please select which platform you are using if self-hosting.

gitlab.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

25.70.4

### Describe the bug

I recently set up a project to use the GitLab Renovate runner. When running the pipeline I get the following error:
`"err": {
         "name": "RequestError",
         "code": "ERR_INVALID_CHAR",
         "message": "Invalid character in header content [\"authorization\"]",`

I have the token set for GITHUB_COM_TOKEN and RENOVATE_TOKEN in the CI/CD vars along with the extra flags vars.
I am using the gitlab shared runners for the pipeline. I thought the issue was maybe the GITHUB token but I have verified it is correct. I have also tried a different project and it returns the same error but for a different npm package. When I look at the package.json file the error happens on the first package listed in the dependencies section.

### Relevant debug logs

<details><summary>Logs</summary>

```
$ renovate $RENOVATE_EXTRA_FLAGS
 INFO: Autodiscovered repositories
       "length": 1,
       "repositories": ["service"]
 INFO: Repository started (repository=service)
       "renovateVersion": "25.70.4"
 INFO: Fallback to renovate.json file as a preset is deprecated, please use a default.json file instead. (repository=service)
 INFO: Fallback to renovate.json file as a preset is deprecated, please use a default.json file instead. (repository=service)
 INFO: Branch created (repository=service, branch=renovate/configure)
       "commit": "fbb627e",
       "onboarding": true
 INFO: Dependency extraction complete (repository=service)
       "baseBranch": "main",
       "stats": {
         "managers": {
           "gitlabci": {"fileCount": 2, "depCount": 13},
           "npm": {"fileCount": 1, "depCount": 67},
           "nvm": {"fileCount": 1, "depCount": 1},
           "terraform": {"fileCount": 4, "depCount": 6}
         },
         "total": {"fileCount": 8, "depCount": 87}
       }
 WARN: Host error (repository=service)
       "hostType": "npm",
       "lookupName": "@babel/runtime",
       "err": {
         "name": "RequestError",
         "code": "ERR_INVALID_CHAR",
         "message": "Invalid character in header content [\"authorization\"]",
         "stack": "RequestError: Invalid character in header content [\"authorization\"]\n    at Request._makeRequest (/usr/src/app/node_modules/got/dist/source/core/index.js:1177:19)\n    at ClientRequest.setHeader (_http_outgoing.js:564:3)\n    at new ClientRequest (_http_client.js:262:14)\n    at request (https.js:326:10)\n    at Request._makeRequest (/usr/src/app/node_modules/got/dist/source/core/index.js:1124:43)\n    at /usr/src/app/node_modules/got/dist/source/core/index.js:328:28",
         "options": {
           "headers": {
             "user-agent": "RenovateBot/25.70.4 (https://github.com/renovatebot/renovate)",
             "authorization": "***********",
             "accept": "application/json",
             "accept-encoding": "gzip, deflate, br"
           },
           "url": "https://registry.npmjs.org/@babel%2Fruntime",
           "username": "",
           "password": "",
           "method": "GET",
           "http2": false
         }
       }
 INFO: External host error causing abort - skipping (repository=service)
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by rarkins** [comment, marked as answer]
Re-run with LOG_LEVEL=debug and pay attention to any mentions of `hostRules` or `npmrc`. It seems that npm authentication is coming from somewhere.

**Reply by tfendt-belong-gg** [reply]
Thanks for the reply! The repo that was scanned has an old .npmrc file in it. This brings up a different question. What is the best way to supply the npmrc information? We typically generate the npmrc file during build time using variables from GitLab.

What is the preferred way to properly set the correct access so we can run renovate against our private npm packages? The docs talk about a config file but I don't have one when using the GitLab pipeline.

**Reply by rarkins** [reply]
https://docs.renovatebot.com/private-npm-modules/

**Reply by tfendt-belong-gg** [reply]
So I removed the npmrc file from the target repo renovate is running against. I am still getting the same error. Is renovate using the RENOVATE_TOKEN by default as the npm token? I am not setting it anywhere but it seems like it is still trying to use something.

**Reply by rarkins** [reply]
No, it doesn't do that. Check the debug logs again for all npmrc mentions and see if anything's left

**Reply by tfendt-belong-gg** [reply]
The only thing I see in the logs is this:
`DEBUG: Resetting npmrc (repository=service)`

I also see this:
```
DEBUG: Env config
       "config": {
         "hostRules": [
           {"hostType": "github", "matchHost": "github.com", "token": "***********"},
           {
             "hostType": "docker",
             "matchHost": "env.ci.runner.short",
             "token": "***********"
           },
           {"hostType": "npm", "token": "***********"}
         ],
```

I am not setting any ENV config so I'm not sure where this is coming from. The only variables I have set on the pipeline are GITHUB_COM_TOKEN, LOG_LEVEL, RENOVATE_EXTRA_FLAGS, and RENOVATE_TOKEN.

The pipeline is just this:
```
include:
  - project: 'renovate-bot/renovate-runner'
    file: '/templates/renovate-dind.gitlab-ci.yml'
```

**Reply by tfendt-belong-gg** [reply]
@rarkins Any thoughts? I followed all the steps here to set up this runner: https://gitlab.com/renovate-bot/renovate-runner
I don't have any extra env settings for npm anywhere yet it is getting added to the hostRules. When I manually add a config file with npm settings I then see two npm rules being set in the combined config...
```json
"config": {
         "hostRules": [
           {"hostType": "npm", "token": "***********"},
           {"hostType": "github", "matchHost": "github.com", "token": "***********"},
           {
             "hostType": "docker",
             "matchHost": "env.ci.runner.short",
             "token": "***********"
           },
           {"hostType": "npm", "token": "***********"}
         ],
```

**Reply by viceice** [reply]
@tfendt-belong-gg You have an env var `NPM_TOKEN` somewhere, which is invalid, so remove it. It will be automatically converted to the seen npm wildcard host rule and applied to all npm registry requests done by renovate.

**Reply by tfendt-belong-gg** [reply]
Thank you! There was a Group variable being inherited called `NPM_TOKEN`. All is well now.
